File: surface/widgets/compose/components/LayoutContainer.kt
// ============================================================
// Clef Surface Compose Widget — LayoutContainer
//
// Compose layout component that implements Clef Surface layout
// kinds (stack, grid, split, overlay, flow, sidebar, center)
// using Compose's Row, Column, LazyVerticalGrid, and Box.
// ============================================================

package clef.surface.compose.components

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// --------------- Layout Kinds ---------------

enum class LayoutKind {
    STACK, ROW, GRID, SPLIT, OVERLAY, FLOW, SIDEBAR, CENTER
}

// --------------- Component ---------------

@Composable
fun LayoutContainer(
    kind: LayoutKind = LayoutKind.STACK,
    gap: Dp = 8.dp,
    padding: Dp = 0.dp,
    gridColumns: Int = 2,
    splitRatio: Float = 0.5f,
    sidebarWidth: Dp = 240.dp,
    sidebarPosition: String = "start",
    crossAxisAlignment: Alignment.Horizontal = Alignment.Start,
    mainAxisAlignment: Arrangement.Vertical = Arrangement.Top,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    when (kind) {
        LayoutKind.STACK -> {
            Column(
                modifier = modifier.padding(padding),
                verticalArrangement = Arrangement.spacedBy(gap),
                horizontalAlignment = crossAxisAlignment,
            ) {
                content()
            }
        }
        LayoutKind.ROW -> {
            Row(
                modifier = modifier.padding(padding),
                horizontalArrangement = Arrangement.spacedBy(gap),
                verticalAlignment = Alignment.Top,
            ) {
                content()
            }
        }
        LayoutKind.GRID -> {
            LazyVerticalGrid(
                columns = GridCells.Fixed(gridColumns),
                modifier = modifier.padding(padding),
                horizontalArrangement = Arrangement.spacedBy(gap),
                verticalArrangement = Arrangement.spacedBy(gap),
            ) {
                // Grid items should be provided via content
            }
        }
        LayoutKind.SPLIT -> {
            Row(
                modifier = modifier
                    .fillMaxWidth()
                    .padding(padding),
                horizontalArrangement = Arrangement.spacedBy(gap),
            ) {
                Box(modifier = Modifier.weight(splitRatio)) {
                    content()
                }
                Box(modifier = Modifier.weight(1f - splitRatio)) {
                    // Second panel content
                }
            }
        }
        LayoutKind.OVERLAY -> {
            Box(
                modifier = modifier.padding(padding),
                contentAlignment = Alignment.Center,
            ) {
                content()
            }
        }
        LayoutKind.FLOW -> {
            // FlowRow is available in Compose Foundation
            Row(
                modifier = modifier.padding(padding),
                horizontalArrangement = Arrangement.spacedBy(gap),
            ) {
                content()
            }
        }
        LayoutKind.SIDEBAR -> {
            Row(
                modifier = modifier
                    .fillMaxWidth()
                    .padding(padding),
                horizontalArrangement = Arrangement.spacedBy(gap),
            ) {
                if (sidebarPosition == "start") {
                    Box(modifier = Modifier.width(sidebarWidth)) { }
                    Box(modifier = Modifier.weight(1f)) { content() }
                } else {
                    Box(modifier = Modifier.weight(1f)) { content() }
                    Box(modifier = Modifier.width(sidebarWidth)) { }
                }
            }
        }
        LayoutKind.CENTER -> {
            Box(
                modifier = modifier
                    .fillMaxSize()
                    .padding(padding),
                contentAlignment = Alignment.Center,
            ) {
                content()
            }
        }
    }
}

File: surface/widgets/nativescript/components/LayoutContainer.ts
// ============================================================
// Clef Surface NativeScript Widget — LayoutContainer
//
// NativeScript layout component that implements Clef Surface
// layout kinds (stack, grid, split, overlay, flow, sidebar,
// center) using NativeScript's StackLayout, GridLayout,
// FlexboxLayout, AbsoluteLayout, and WrapLayout.
// ============================================================

import {
  StackLayout,
  GridLayout,
  FlexboxLayout,
  WrapLayout,
  AbsoluteLayout,
  DockLayout,
  View,
} from '@nativescript/core';

// --------------- Layout Kinds ---------------

export type LayoutKind = 'stack' | 'row' | 'grid' | 'split' | 'overlay' | 'flow' | 'sidebar' | 'center';

// --------------- Props ---------------

export interface LayoutContainerProps {
  kind?: LayoutKind;
  gap?: number;
  padding?: number;
  gridColumns?: number;
  splitRatio?: number;
  sidebarWidth?: number;
  sidebarPosition?: 'start' | 'end';
}

// --------------- Component ---------------

export function createLayoutContainer(props: LayoutContainerProps = {}): View {
  const {
    kind = 'stack',
    gap = 8,
    padding = 0,
    gridColumns = 2,
    splitRatio = 0.5,
    sidebarWidth = 240,
    sidebarPosition = 'start',
  } = props;

  switch (kind) {
    case 'stack': {
      const layout = new StackLayout();
      layout.className = 'clef-layout clef-layout-stack';
      layout.padding = padding;
      return layout;
    }
    case 'row': {
      const layout = new FlexboxLayout();
      layout.className = 'clef-layout clef-layout-row';
      layout.flexDirection = 'row';
      layout.padding = padding;
      return layout;
    }
    case 'grid': {
      const layout = new GridLayout();
      layout.className = 'clef-layout clef-layout-grid';
      const cols = Array(gridColumns).fill('*').join(', ');
      layout.columns = cols;
      layout.padding = padding;
      return layout;
    }
    case 'split': {
      const layout = new GridLayout();
      layout.className = 'clef-layout clef-layout-split';
      const leftStar = Math.round(splitRatio * 10);
      const rightStar = 10 - leftStar;
      layout.columns = `${leftStar}*, ${rightStar}*`;
      layout.padding = padding;
      return layout;
    }
    case 'overlay': {
      const layout = new AbsoluteLayout();
      layout.className = 'clef-layout clef-layout-overlay';
      layout.padding = padding;
      return layout;
    }
    case 'flow': {
      const layout = new WrapLayout();
      layout.className = 'clef-layout clef-layout-flow';
      layout.orientation = 'horizontal';
      layout.padding = padding;
      return layout;
    }
    case 'sidebar': {
      const layout = new GridLayout();
      layout.className = 'clef-layout clef-layout-sidebar';
      if (sidebarPosition === 'start') {
        layout.columns = `${sidebarWidth}, *`;
      } else {
        layout.columns = `*, ${sidebarWidth}`;
      }
      layout.padding = padding;
      return layout;
    }
    case 'center': {
      const layout = new FlexboxLayout();
      layout.className = 'clef-layout clef-layout-center';
      layout.justifyContent = 'center';
      layout.alignItems = 'center';
      layout.padding = padding;
      return layout;
    }
    default: {
      const layout = new StackLayout();
      layout.className = 'clef-layout';
      layout.padding = padding;
      return layout;
    }
  }
}

createLayoutContainer.displayName = 'LayoutContainer';
export default createLayoutContainer;
